File: template/perl/github_update.pl
#!/usr/bin/env perl
# github_update.pl
# to run: perl github_update.pl [--debug]

use strict;
use warnings;
use Getopt::Long;
use File::Basename;

my $DEBUG = 0;

sub debug_print {
    my ($message) = @_;
    print "DEBUG: $message\n" if $DEBUG;
}

sub run_git_command {
    my ($command) = @_;
    debug_print("Running command: $command");
    my $output = `$command 2>&1`;
    my $exit_code = $? >> 8;
    debug_print("Command output: $output");
    debug_print("Command exit code: $exit_code");
    return ($output, $exit_code);
}

sub check_for_uncommitted_changes {
    print "Checking for uncommitted changes...\n";
    my ($status_output, $exit_code) = run_git_command("git status --porcelain");
    if ($status_output) {
        print "Uncommitted changes found.\n";
    } else {
        print "No uncommitted changes.\n";
    }
    return $status_output ? 1 : 0;
}

sub stash_changes {
    print "Stashing local changes...\n";
    my ($output, $exit_code) = run_git_command("git stash");
    if ($output =~ /No local changes to save/) {
        print "No changes to stash.\n";
    } elsif ($exit_code != 0) {
        print "Error stashing changes: $output\n";
        exit 1;
    } else {
        print "Changes stashed successfully.\n";
    }
}

sub pop_stashed_changes {
    print "Applying stashed changes...\n";
    my ($output, $exit_code) = run_git_command("git stash pop");
    if ($output =~ /No stash entries found/) {
        print "No stashed changes to apply.\n";
    } elsif ($exit_code != 0 && $output !~ /CONFLICT/) {
        print "Error applying stashed changes: $output\n";
        exit 1;
    } elsif ($output =~ /CONFLICT/) {
        print "Conflicts occurred while applying stashed changes. Please resolve manually.\n";
    } else {
        print "Stashed changes applied successfully.\n";
    }
}

sub fetch_remote {
    print "Fetching remote changes...\n";
    my ($output, $exit_code) = run_git_command("git fetch");
    if ($exit_code != 0) {
        print "Error fetching remote: $output\n";
        exit 1;
    }
    print "Remote changes fetched successfully.\n";
}

sub check_for_diverged_history {
    print "Checking for diverged history...\n";
    my ($local_commit, $exit_code) = run_git_command("git rev-parse HEAD");
    my ($remote_commit, $exit_code2) = run_git_command("git rev-parse \@{u}");
    chomp($local_commit, $remote_commit);
    my $diverged = $local_commit ne $remote_commit;
    if ($diverged) {
        print "Local and remote histories have diverged.\n";
    } else {
        print "Local and remote histories are in sync.\n";
    }
    return $diverged;
}

sub merge_remote_changes {
    print "Merging remote changes...\n";
    my ($output, $exit_code) = run_git_command("git merge origin/main");
    if ($output =~ /CONFLICT/) {
        print "Merge conflict detected. Please resolve conflicts manually.\n";
        exit 1;
    } elsif ($exit_code != 0) {
        print "Error merging changes: $output\n";
        exit 1;
    } else {
        print "Remote changes merged successfully.\n";
    }
}

sub push_changes {
    print "Pushing local changes to remote...\n";
    my ($output, $exit_code) = run_git_command("git push");
    if ($exit_code != 0) {
        print "Error pushing changes: $output\n";
        exit 1;
    }
    print "Local changes pushed successfully.\n";
}

sub github_update {
    my $stashed = 0;
    if (check_for_uncommitted_changes()) {
        stash_changes();
        $stashed = 1;
    }

    fetch_remote();

    if (check_for_diverged_history()) {
        merge_remote_changes();
    }

    push_changes();

    if ($stashed) {
        pop_stashed_changes();
    }

    print "GitHub repository updated successfully.\n";
}

GetOptions("debug" => \$DEBUG) or die("Error in command line arguments\n");

github_update();

File: template/php/github_update.php
<?php
// github_update.php
// to run: php github_update.php [--debug]

$DEBUG = false;

function debug_print($message) {
	global $DEBUG;
	if ($DEBUG) {
		echo "DEBUG: $message\n";
	}
}

function run_git_command($command) {
	debug_print("Running command: $command");
	$output = [];
	$return_var = 0;
	exec($command . " 2>&1", $output, $return_var);
	$output_str = implode("\n", $output);
	debug_print("Command output: $output_str");
	return [$output_str, $return_var];
}

function check_for_uncommitted_changes() {
	echo "Checking for uncommitted changes...\n";
	list($status_output, $return_var) = run_git_command("git status --porcelain");
	if ($status_output) {
		echo "Uncommitted changes found.\n";
	} else {
		echo "No uncommitted changes.\n";
	}
	return !empty($status_output);
}

function stash_changes() {
	echo "Stashing local changes...\n";
	list($output, $return_var) = run_git_command("git stash");
	if (strpos($output, "No local changes to save") !== false) {
		echo "No changes to stash.\n";
	} elseif ($return_var !== 0) {
		echo "Error stashing changes: $output\n";
		exit(1);
	} else {
		echo "Changes stashed successfully.\n";
	}
}

function pop_stashed_changes() {
	echo "Applying stashed changes...\n";
	list($output, $return_var) = run_git_command("git stash pop");
	if (strpos($output, "No stash entries found") !== false) {
		echo "No stashed changes to apply.\n";
	} elseif ($return_var !== 0 && strpos($output, "CONFLICT") === false) {
		echo "Error applying stashed changes: $output\n";
		exit(1);
	} elseif (strpos($output, "CONFLICT") !== false) {
		echo "Conflicts occurred while applying stashed changes. Please resolve manually.\n";
	} else {
		echo "Stashed changes applied successfully.\n";
	}
}

function fetch_remote() {
	echo "Fetching remote changes...\n";
	list($output, $return_var) = run_git_command("git fetch");
	if ($return_var !== 0) {
		echo "Error fetching remote: $output\n";
		exit(1);
	}
	echo "Remote changes fetched successfully.\n";
}

function check_for_diverged_history() {
	echo "Checking for diverged history...\n";
	list($local_commit, $return_var) = run_git_command("git rev-parse HEAD");
	list($remote_commit, $return_var) = run_git_command("git rev-parse @{u}");
	$diverged = trim($local_commit) !== trim($remote_commit);
	if ($diverged) {
		echo "Local and remote histories have diverged.\n";
	} else {
		echo "Local and remote histories are in sync.\n";
	}
	return $diverged;
}

function merge_remote_changes() {
	echo "Merging remote changes...\n";
	list($output, $return_var) = run_git_command("git merge origin/main");
	if (strpos($output, "CONFLICT") !== false) {
		echo "Merge conflict detected. Please resolve conflicts manually.\n";
		exit(1);
	} elseif ($return_var !== 0) {
		echo "Error merging changes: $output\n";
		exit(1);
	} else {
		echo "Remote changes merged successfully.\n";
	}
}

function push_changes() {
	echo "Pushing local changes to remote...\n";
	list($output, $return_var) = run_git_command("git push");
	if ($return_var !== 0) {
		echo "Error pushing changes: $output\n";
		exit(1);
	}
	echo "Local changes pushed successfully.\n";
}

function github_update() {
	// Check for uncommitted changes
	if (check_for_uncommitted_changes()) {
		stash_changes();
		$stashed = true;
	} else {
		$stashed = false;
	}

	// Fetch remote changes
	fetch_remote();

	// Check for diverged history
	if (check_for_diverged_history()) {
		merge_remote_changes();
	}

	// Push local changes
	push_changes();

	// Apply stashed changes if any
	if ($stashed) {
		pop_stashed_changes();
	}

	echo "GitHub repository updated successfully.\n";
}

// Parse command line arguments
$options = getopt("", ["debug"]);
$DEBUG = isset($options["debug"]);

github_update();